Test Failure (20180112-0701): ConcurrentRxTestServlet.testNoNewMethods expected 104 but was 117

A test case that we wrote to identify when new methods are added to CompletableFuture (which need to be properly accounted for in code) has raised an error in the Java 9 builds letting us know that Java 9 has added 13 new methods to CompletableFuture that will need to be investigated and added to our implementation:

public CompletableFuture<T> completeAsync​(Supplier<? extends T> supplier)
public CompletableFuture<T> completeAsync​(Supplier<? extends T> supplier, Executor executor)
public CompletableFuture<T> completeOnTimeout​(T value, long timeout, TimeUnit unit)
public CompletableFuture<T> copy​()
public Executor defaultExecutor​()
public CompletionStage<T> minimalCompletionStage​()
public <U> CompletableFuture<U> newIncompleteFuture​()
public CompletableFuture<T> orTimeout​(long timeout, TimeUnit unit)

public static <U> CompletionStage<U> completedStage​(U value)
public static Executor delayedExecutor​(long delay, TimeUnit unit)
public static Executor delayedExecutor​(long delay, TimeUnit unit, Executor executor)
public static <U> CompletableFuture<U> failedFuture​(Throwable ex)
public static <U> CompletionStage<U> failedStage​(Throwable ex)

For reference, here is how the test case reports it:
junit.framework.AssertionFailedError: 2018-01-12-22:40:07:689 The response did not contain [SUCCESS].  Full output is:
	ERROR: Caught exception attempting to call test method testNoNewMethods on servlet web.ConcurrentRxTestServlet
	java.lang.AssertionError: Methods have been added to CompletableFuture which need to be properly implemented on wrapper class ManagedCompletableFuture expected:<104> but was:<117>
	at web.ConcurrentRxTestServlet.testNoNewMethods(ConcurrentRxTestServlet.java:1554)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@9-internal/Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@9-internal/NativeMethodAccessorImpl.java:62)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@9-internal/DelegatingMethodAccessorImpl.java:43)
	at componenttest.app.FATServlet.doGet(FATServlet.java:71)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:686)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:791)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1255)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:743)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:440)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1154)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1004)
	at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:76)
	at com.ibm.ws.webcontainer40.servlet.CacheServletWrapper40.handleRequest(CacheServletWrapper40.java:83)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:926)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:989)
	at com.ibm.ws.http.dispatcher.internal.ch

--- a/dev/com.ibm.websphere.appserver.concurrentRX-1.0/com.ibm.websphere.appserver.concurrentRX-1.0.feature
+++ b/dev/com.ibm.websphere.appserver.concurrentRX-1.0/com.ibm.websphere.appserver.concurrentRX-1.0.feature
@@ -5,5 +5,6 @@ IBM-Provision-Capability: osgi.identity; filter:="(&(type=osgi.subsystem.feature
  osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=com.ibm.websphere.appserver.jaxrs-2.1))"
 IBM-Install-Policy: when-satisfied
 -bundles=com.ibm.ws.concurrent.rx
+# Do not release this feature until ManagedCompletableFuture has been properly rebased for Java 9 (see TODO comments in that file)
 kind=beta
 edition=core

--- a/dev/com.ibm.ws.concurrent.rx/src/com/ibm/websphere/concurrent/rx/ManagedCompletableFuture.java
+++ b/dev/com.ibm.ws.concurrent.rx/src/com/ibm/websphere/concurrent/rx/ManagedCompletableFuture.java
@@ -69,6 +69,11 @@ import com.ibm.wsspi.threadcontext.WSContextService;
  * one for the subclass itself, which includes the override logic, which delegates to another CompletableFuture
  * instance that does the actual work.
  *
+ * TODO Before releasing any code that uses this, this implementation needs to be rebased onto Java 9,
+ * taking advantage of plugin points for defaultExecutor/newIncompleteFuture, which will allow for a more
+ * optimal implementation. Java 9 methods are marked with TODO, however, all methods should be impacted
+ * by this.
+ *
  * @param <T>
  */
 public class ManagedCompletableFuture<T> extends CompletableFuture<T> {
@@ -218,6 +223,47 @@ public class ManagedCompletableFuture<T> extends CompletableFuture<T> {
                         CompletableFuture.completedFuture(value), //
                         AccessController.doPrivileged(getDefaultManagedExecutorAction), //
                         null);
+    }
+
+    /**
+     * Replaces CompletableFuture.completedStage(value) with an implementation that switches the
+     * default asynchronous execution facility to be the default managed executor.
+     *
+     * @param value result of the completed future
+     * @return completed completion stage where the default managed executor is the default asynchronous execution facility.
+     */
+    public static <U> CompletionStage<U> completedStage(U value) {
+        throw new UnsupportedOperationException(); // TODO implement when rebasing on Java 9
+    }
+
+    public static Executor delayedExecutor​(long delay, TimeUnit unit) {
+        throw new UnsupportedOperationException(); // TODO implement when rebasing on Java 9
+    }
+
+    public static Executor delayedExecutor​(long delay, TimeUnit unit, Executor executor) {
+        throw new UnsupportedOperationException(); // TODO implement when rebasing on Java 9
+    }
+
+    /**
+     * Replaces CompletableFuture.failedFuture(value) with an implementation that switches the
+     * default asynchronous execution facility to be the default managed executor.
+     *
+     * @param value result of the completed future
+     * @return completed completable future where the default managed executor is the default asynchronous execution facility.
+     */
+    public static <U> ManagedCompletableFuture<U> failedFuture(U value) {
+        throw new UnsupportedOperationException(); // TODO implement when rebasing on Java 9
+    }
+
+    /**
+     * Replaces CompletableFuture.failedStage(value) with an implementation that switches the
+     * default asynchronous execution facility to be the default managed executor.
+     *
+     * @param value result of the completed future
+     * @return completed completion stage where the default managed executor is the default asynchronous execution facility.
+     */
+    public static <U> CompletionStage<U> failedStage(U value) {
+        throw new UnsupportedOperationException(); // TODO implement when rebasing on Java 9
     }
 
     /**
@@ -473,6 +519,22 @@ public class ManagedCompletableFuture<T> extends CompletableFuture<T> {
     }
 
     /**
+     * @see java.util.concurrent.CompletableFuture#completeAsync(Supplier)
+     */
+    // TODO Java 9 @Override
+    public CompletableFuture<T> completeAsync​(Supplier<? extends T> supplier) {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * @see java.util.concurrent.CompletableFuture#completeAsync(Supplier, Executor)
+     */
+    // TODO Java 9 @Override
+    public CompletableFuture<T> completeAsync​(Supplier<? extends T> supplier, Executor executor) {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
      * @see java.util.concurrent.CompletableFuture#completeExceptionally(java.lang.Throwable)
      */
     @Override
@@ -487,6 +549,30 @@ public class ManagedCompletableFuture<T> extends CompletableFuture<T> {
         }
 
         return completedByThisMethod;
+    }
+
+    /**
+     * @see java.util.concurrent.CompletableFuture#completeOnTimeout(Object, long, TimeUnit)
+     */
+    // TODO Java 9 @Override
+    public CompletableFuture<T> completeOnTimeout​(T value, long timeout, TimeUnit unit) {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * @see java.util.concurrent.CompletableFuture#copy()
+     */
+    // TODO Java 9 @Override
+    public CompletableFuture<T> copy​() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * @see java.util.concurrent.CompletableFuture#defaultExecutor()
+     */
+    // TODO Java 9 @Override
+    public Executor defaultExecutor​() {
+        return defaultExecutor;
     }
 
     /**
@@ -634,6 +720,22 @@ public class ManagedCompletableFuture<T> extends CompletableFuture<T> {
     }
 
     /**
+     * @see java.util.concurrent.CompletableFuture#minimalCompletionStage()
+     */
+    // TODO Java 9 @Override
+    public CompletionStage<T> minimalCompletionStage() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * @see java.util.concurrent.CompletableFuture#newIncompleteFuture()
+     */
+    // TODO Java 9 @Override
+    public CompletableFuture<T> newIncompleteFuture() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
      * @see java.util.concurrent.CompletableFuture#obtrudeException(java.lang.Throwable)
      */
     @Override
@@ -669,6 +771,14 @@ public class ManagedCompletableFuture<T> extends CompletableFuture<T> {
             if (future != null)
                 future.cancel(true);
         }
+    }
+
+    /**
+     * @see java.util.concurrent.CompletableFuture#orTimeout(long, TimeUnit)
+     */
+    // TODO Java 9 @Override
+    public CompletableFuture<T> orTimeout(long timeout, TimeUnit unit) {
+        throw new UnsupportedOperationException();
     }
 
     /**

--- a/dev/com.ibm.ws.concurrent_fat_rx/test-applications/concurrentrxfat/src/web/ConcurrentRxTestServlet.java
+++ b/dev/com.ibm.ws.concurrent_fat_rx/test-applications/concurrentrxfat/src/web/ConcurrentRxTestServlet.java
@@ -1610,9 +1610,11 @@ public class ConcurrentRxTestServlet extends FATServlet {
      */
     @Test
     public void testNoNewMethods() throws Exception {
-        assertEquals("Methods have been added to CompletableFuture which need to be properly implemented on wrapper class ManagedCompletableFuture",
-                     104, // WARNING: do not update this value unless you have properly implemented the new methods on ManagedCompletableFuture!
-                     CompletableFuture.class.getMethods().length);
+        int methodCount = CompletableFuture.class.getMethods().length;
+        assertTrue("Methods have been added to CompletableFuture which need to be properly implemented on wrapper class ManagedCompletableFuture. " +
+                   "Expected 117 (Java 9) or 104 (Java 8). Found " + methodCount,
+                   // WARNING: do not update these values unless you have properly implemented (or added code to reject) the new methods on ManagedCompletableFuture!
+                   methodCount == 117 || methodCount == 104);
     }
 
     /**

